subprojects/core/src/main/java/org/gradle/api/internal/project/ProjectIdentity.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.project;
+
+import org.gradle.api.artifacts.component.BuildIdentifier;
+import org.gradle.util.Path;
+
+/**
+ * Identifies a single project within the build and the build tree.
+ */
+public final class ProjectIdentity {
+
+    private final BuildIdentifier buildIdentifier;
+    private final Path buildTreePath;
+    private final Path projectPath;
+    private final String projectName;
+
+    public ProjectIdentity(
+        BuildIdentifier buildIdentifier,
+        Path buildTreePath,
+        Path projectPath,
+        String projectName
+    ) {
+        this.buildIdentifier = buildIdentifier;
+        this.buildTreePath = buildTreePath;
+        this.projectPath = projectPath;
+        this.projectName = projectName;
+    }
+
+    /**
+     * The identity of the owning build.
+     */
+    public BuildIdentifier getBuildIdentifier() {
```

Comment:
I'm slightly sad this general core concept has to carry something more dependency management related.

I guess, in the ideal world we would have a `BuildIdentity` class as well, and this is what `ProjectIdentity` would refer to, and the DM logic would rely on those to derive its `BuildIdentifier` instances.

Reply:
`BuildIdentifier` is really a wrapper around a single `Path`. I think this class would eventually just be able to hold a `Path` to represent the build instead of a `BuildIdentifier`. 

However, `BuildIdentifier` has two deprecated methods -- `getName` and `isCurrentBuild` -- which will be removed in 9.0

I think once that time comes around, we can simplify this class to only have a `Path` to represent the build and not the whole `BuildIdentifier` instance.

Reply:
Part of my comment was related to the fact that ideally the class representing a build identity should not live in the "artifacts" related package. Also, it does not have to be public API, while BuildIdentifier is already.

Reply:
Yeah I definitely agree, though if you look at `BuildState` you will see it already uses this `BuildIdentifier`. 

I just think its worth waiting till we simplify `BuildIdentifier` to avoid rework when changes are made there